### ISSUE TITLE
selection dimensions in visual mode

### ISSUE THREAD

**Issue by mkaminsky** [t9md/atom-vim-mode-plus#47]
First, thank you for the package--it seems to be working very nicely.

In Atom, when selecting text, the status bar at the bottom shows the line and character count of the selection in parenthesis.  For example: `(1,7)` to indicate that the current selection is 1 line tall and contains 7 characters.  In Visual Character mode, the dimensions seem accurate, but in vim-mode-plus's Visual Line and Visual Block modes, the line count appears to be off (in two different ways).

Using <kbd>V</kbd> to select the current line gives a line count of 2.  Selecting two lines in Visual Line mode gives a line count of 3, and so forth.  The character count appears correct, assuming that the newline characters count as 1, but the line count is one number too high.

Using <kbd>Ctrl</kbd>+<kbd>v</kbd> to enter Visual Block mode, though, always produces a line count of 1, no matter how many lines the block selection spans.  The second number in the tuple, the character count, is equal to the width of the selection, not the total number of characters.

The behavior in Visual Line mode seems reasonable, except for the line count being one number too high.  The behavior in Visual Block mode also seems reasonable, except for the line count always being one.  Having the character count be the width of the selection (instead of total number of characters) in this mode seems right to me, and mimics Vim's behavior for block selections which is represented as `height x width`.

Thanks!

**Comment by t9md** [comment]
visual block is just using multi-cursor, multi-selection under the hood.
Your report seem to be correct.
To improve this, provide vim-mode-plus aware cursor, selection indicator on StatusBar.
Technically possible, not difficult but priority is very low for me for now.
Thanks for feedback.

**Comment by ypresto** [comment]
This is problematic with plugins and custom commands deal with current selected **rows**.

**Comment by t9md** [comment]
@ypresto
Give me real situation as example.

**Comment by ypresto** [comment]
This is my script to append some character (like ':', ';', ',') to each selected lines.

```coffee
appendString = (event, string) ->
  editor = atom.workspace.getActiveTextEditor()
  rows = new Set()
  # TODO: Use map and flatten
  editor.getSelectedBufferRanges().forEach (range) ->
    range.getRows().forEach (row) -> rows.add(row)
  editor.transact ->
    rows.forEach (row) ->
      point = [row, Number.MAX_SAFE_INTEGER]
      editor.setTextInBufferRange([point, point], string)
```

**Comment by ypresto** [comment]
With visual line mode `rows` contains one more line than actually selected.
So this command results:

```
foo
bar
```

to be

```
foo:
bar:
:
```

**Comment by t9md** [comment]
What's the problem please describe the problematic situation you think.

**Comment by t9md** [comment]
Oh ok

**Comment by t9md** [comment]
The issue you described is different from this issue's subject.
When you select line in Atom, cursor is set from column 0 of current line to **next line** of column 0.
Its Atom's way to select line. You can check by range.toString().
In your code you can solve problem by using selection.getRowRanges which return startRow and endRow of actually selected row.

**Comment by ypresto** [comment]
Oh really...
I was misunderstanding problem.
I tested 'editor:indent-selected-rows' and it works well with visual line mode.
Sorry!

**Comment by ypresto** [comment]
Selection.getBufferRowRange() does the trick, thanks!

**Comment by t9md** [comment]
Yeah, Atom's way of line selection is counter intuitive. I have to always use selection.getRowRange.
